feat(scan): introduce a client to manage scans via REST API

ostridm opened the pull request:
closes #44

ostridm committed:
feat(scan): introduce a client to manage scans via REST API

closes #44

derevnjuk commented on src/SecTester.Bus/Extensions/MessageSerializerExtensions.cs:
```diff
@@ -0,0 +1,14 @@
+
+public static class MessageSerializerExtensions
+{
+  public static HttpContent SerializeJsonContent<T>(this MessageSerializer messageSerializer, T data)
```
The method name confuses: 
* You serialize data passed by the first argument, but not the content. 
* The content is created by calling this method but not serialized.
* Why does the serializer responsible to create HttpContent?

derevnjuk commented on src/SecTester.Bus/Extensions/MessageSerializerExtensions.cs:
```diff
@@ -0,0 +1,14 @@
+{
+  public static HttpContent SerializeJsonContent<T>(this MessageSerializer messageSerializer, T data)
+  {
+    return new StringContent(messageSerializer.Serialize(data), Encoding.UTF8, "application/json");
```
Why don't create `JsonContent` instead?

ostridm replied:
`JsonContent` is missing when the target is `netstadard`. It appears for `.net 5` and above.

derevnjuk commented on src/SecTester.Scan/Commands/CreateScan.cs:
```diff
@@ -0,0 +1,13 @@
+
+public record CreateScan : HttpRequest<Identifiable<string>>
+{
+  public CreateScan(HttpContent body, int? ttl = null)
```
Well, does it mean I can pass `multipart/form-data` with any data to create a scan? It looks like you have to accept DTO to avoid ambiguity

derevnjuk commented on src/SecTester.Scan/Commands/DeleteScan.cs:
```diff
@@ -0,0 +1,12 @@
+
+public record DeleteScan : HttpRequest<Unit>
+{
+  public DeleteScan(string id, int? ttl = default)
```
In some place you use the default, but in some null. You have to decide. Also, I don't see any reason to expose TTL for these commands.

derevnjuk commented on src/SecTester.Scan/DefaultScans.cs:
```diff
@@ -0,0 +1,111 @@
+    var fileNameContent = new StringContent(options.FileName);
+    
+    var requestContent = new MultipartFormDataContent();
+    requestContent.Add(jsonFileContent, "file");
+    requestContent.Add(fileNameContent, "filename");
```
Since when does the filename is a separate field?
```suggestion
    requestContent.Add(jsonFileContent, "file", fileNameContent);
```

derevnjuk commented on src/SecTester.Scan/DefaultScans.cs:
```diff
@@ -0,0 +1,111 @@
+    requestContent.Add(jsonFileContent, "file");
+    requestContent.Add(fileNameContent, "filename");
+    
+    var result =  await SendCommand(new UploadHar(requestContent, options.Discard));
```
```suggestion
    var result = await SendCommand(new UploadHar(requestContent, options.Discard));
```

derevnjuk commented on src/SecTester.Scan/DefaultScans.cs:
```diff
@@ -0,0 +1,111 @@
+    return result.Id;
+  }
+
+  public async Task<IEnumerable<Issue>> ListIssues(string id)
+  {
+    return await SendCommand(new ListIssues(id));
+  }
```
```suggestion
  public Task<IEnumerable<Issue>> ListIssues(string id)
  {
    return SendCommand(new ListIssues(id));
  }
```

derevnjuk commented on src/SecTester.Scan/DefaultScans.cs:
```diff
@@ -0,0 +1,111 @@
+
+  private async Task<T> SendCommand<T>(Command<T> command)
+  {
+    var result = await this._commandDispatcher.Execute(command);
```
Some unit test frameworks (like xUnit) have their own synchronization context. So, you should use `ConfigureAwait(false)` everywhere. To make this easier, you can add `configure_await_analysis_mode = library` to `.editorconfig`

derevnjuk commented on src/SecTester.Scan/DefaultScans.cs:
```diff
@@ -0,0 +1,111 @@
+
+namespace SecTester.Scan;
+
+internal class DefaultScans : Scans
```
It is not clear to me, why commands are public, but this one is internal. It should be the opposite.

derevnjuk commented on src/SecTester.Scan/CI/DefaultCiDiscovery.cs:
```diff
@@ -0,0 +1,11 @@
+// TODO (dmitry.osrikov@brightsec.com): rework using 'ci-info' compatible implementation
+internal class DefaultCiDiscovery : CiDiscovery
+{
+  public CiServer Server { get; } = CiServer.Unknown;
```
```suggestion
  public CiServer? Server { get; };
```

derevnjuk commented on src/SecTester.Scan/CI/DefaultCiDiscovery.cs:
```diff
@@ -0,0 +1,11 @@
+{
+  public CiServer Server { get; } = CiServer.Unknown;
+
+  public bool IsCi => Server != CiServer.Unknown;
```
```suggestion
  public bool IsCi => Server != null;
```

derevnjuk commented on src/SecTester.Scan/CI/DefaultCiDiscovery.cs:
```diff
@@ -0,0 +1,11 @@
+
+  public bool IsCi => Server != CiServer.Unknown;
+
+  public bool IsPr { get; }
```
Remove unused property

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -0,0 +1,325 @@
+  public void Dispose()
+  {
+    _ciDiscovery.ClearSubstitute();
+    _commandDispatcher.ClearSubstitute();
```
```suggestion
    _commandDispatcher.ClearSubstitute();
    GC.SuppressFinalize(this);
```

derevnjuk commented on test/SecTester.Scan.Tests/Extensions/TaskExtensions.cs:
```diff
@@ -0,0 +1,9 @@
+
+public static class TaskExtensions
+{
+  public static TResult? GetSync<T, TResult>(this T? t, Func<T, Task<TResult>> func)
```
Please follow conventions

```suggestion
  public static TResult? ContinueWithSycn<T, TResult>(this T? t, Func<T, Task<TResult>> continuationFunction)
```

derevnjuk commented on test/SecTester.Scan.Tests/Extensions/TaskExtensions.cs:
```diff
@@ -0,0 +1,9 @@
+namespace SecTester.Scan.Tests.Extensions;
+
+public static class TaskExtensions
```
This extension must be internal

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
Well, the expectation is too complex. You just need to make sure that one of an item of `MultipartFormDataContent` contains a field with a `jsonFileContent`

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
It is not clear why you test `UploadHar` here. How does it relate to Scans? You just need to make sure that the command dispatcher receives a proper instance of `UploadHar`

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```suggestion
  public async Task UploadHar_ResultIsNull_ThrowError()
```

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```suggestion
  public async Task GetScan_ResultIsNull_ThrowError()
```

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```suggestion
  public async Task ListIssues_ResultIsNull_ThrowError()
```

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  [Fact]
+  public async Task ListIssues_ReturnAListOfIssues()
```
```suggestion
  public async Task ListIssues_ReturnListOfIssues()
```

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  [Fact]
+  public async Task CreateScan_WhenResultIsNull_ThrowError()
```
```suggestion
  public async Task CreateScan_ResultIsNull_ThrowError()
```

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -0,0 +1,325 @@
+    // assert
+    await _commandDispatcher.Received(1)
+      .Execute(Arg.Is<CreateScan>(x => x.Url.EndsWith("/api/v1/scans") && x.Method == HttpMethod.Post &&
+                                       x.ExpectReply &&
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -0,0 +1,325 @@
+  {
+    // arrange
+    _commandDispatcher.Execute(Arg.Any<CreateScan>())
+      .Returns(Task.FromResult<Identifiable<string>?>(new Identifiable<string>("id")));
```
The test data must be as close as possible to real

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -0,0 +1,325 @@
+    Smart = true,
+    Tests = new[] { TestType.Csrf, TestType.Jwt },
+    DiscoveryTypes = new[] { Discovery.Crawler },
+    FileId = "fileId",
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -0,0 +1,325 @@
+    FileId = "fileId",
+    HostsFilter = new[] { "example.com" },
+    PoolSize = 2,
+    ProjectId = "projectId",
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -0,0 +1,325 @@
+  private readonly ScanConfig _scanConfig = new("scan1")
+  {
+    Module = Module.Dast,
+    Repeaters = new[] { "repeater1" },
```
ditto

derevnjuk commented on test/SecTester.Bus.Tests/Extensions/MessageSerializerExtensionsTests.cs:
```diff
@@ -0,0 +1,32 @@
+  public void SerializeJsonContent_ReturnsContentWithJsonMediaType()
+  {
+    
+    // arrange
```
```suggestion
  {
    // arrange
```

derevnjuk reviewed (changes requested):
There are a lot of styling and formatting issues (including CRLF), make sure that dotnet tools and husky work as expected. If not, use `dotnet format` manually

ostridm committed:
feat(scan): introduce a client to manage scans via REST API

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031204383
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031258223
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031267711
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031290793
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031291418
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031291500
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031291639
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031292181
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031289660
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031289568
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031289447
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031289265
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031288956
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031244320
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031227859
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031246337
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031246007
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031245764
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031272173
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031225330
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031197621
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031201963
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031225330
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031227044
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031234321
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031282332
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031287911
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031288281
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031288510
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031288628
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031288682
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031288760
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1031289759

closes #44

derevnjuk commented on src/SecTester.Scan/Commands/CreateScan.cs:
```diff
@@ -0,0 +1,42 @@
+    Configuration configuration)
+    : base("/api/v1/scans", HttpMethod.Post)
+  {
+    var payload = new
```
I don't like that you delegate creating a DTO to the command.

derevnjuk commented on test/SecTester.Scan.Tests/Commands/DeleteScanTests.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace SecTester.Scan.Tests.Commands;
+
+public class DeleteScanTests : ScanFixture
```
We have already discussed, the constructor must be tested only in the case of presence logic.

derevnjuk commented on test/SecTester.Scan.Tests/Commands/GetScanTests.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace SecTester.Scan.Tests.Commands;
+
+public class GetScanTests : ScanFixture
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Commands/ListIssuesTests.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace SecTester.Scan.Tests.Commands;
+
+public class ListIssuesTests : ScanFixture
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Commands/StopScanTests.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace SecTester.Scan.Tests.Commands;
+
+public class StopScanTests : ScanFixture
```
ditto

derevnjuk commented on src/SecTester.Scan/Content/DefaultHttpContentFactory.cs:
```diff
@@ -0,0 +1,28 @@
+    return new StringContent(_messageSerializer.Serialize(data), Encoding.UTF8, "application/json");
+  }
+
+  public virtual HttpContent CreateHarContent(UploadHarOptions options)
```
Well, I don't see any reason to introduce a method that will be used only once in the particular case

derevnjuk commented on src/SecTester.Scan/Content/DefaultHttpContentFactory.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace SecTester.Scan.Content;
+
+public class DefaultHttpContentFactory : HttpContentFactory
```
Actually, I don't see any reason to have such a class. Let's discuss this on our call

derevnjuk commented on src/SecTester.Scan/Commands/CreateScan.cs:
```diff
@@ -0,0 +1,42 @@
+
+internal record CreateScan : HttpRequest<Identifiable<string>>
+{
+  public CreateScan(ScanConfig config, HttpContentFactory httpContentFactory, CiDiscovery ciDiscovery,
+    Configuration configuration)
+    : base("/api/v1/scans", HttpMethod.Post)
```
The commands should not depend on any kind of service, it makes them unusable. According to the concept, they must be immutable simple POJO/DTO

derevnjuk commented on test/SecTester.Scan.Tests/Commands/UploadHarTests.cs:
```diff
@@ -0,0 +1,40 @@
+    var command = new UploadHar(options, HttpContentFactory);
+
+    // assert
+    command.Body.Should().Be(content);
+    command.Url.Should().Be("/api/v1/files");
+    command.Method.Should().Be(HttpMethod.Post);
+    command.ExpectReply.Should().BeTrue();
+    command.Params.Should().BeNull();
```
You shiould use a partial matcher instead

```csharp
command.Should(). BeEquivalentTo(new {})
```

derevnjuk commented on test/SecTester.Scan.Tests/Commands/StopScanTests.cs:
```diff
@@ -0,0 +1,19 @@
+    var command = new StopScan(ScanId);
+
+    // assert
+    command.Url.Should().Be($"/api/v1/scans/{ScanId}/stop");
+    command.Method.Should().Be(HttpMethod.Get);
+    command.ExpectReply.Should().BeFalse();
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Commands/ListIssuesTests.cs:
```diff
@@ -0,0 +1,19 @@
+    var command = new ListIssues(ScanId);
+
+    // assert
+    command.Url.Should().Be($"/api/v1/scans/{ScanId}/issues");
+    command.Method.Should().Be(HttpMethod.Get);
+    command.ExpectReply.Should().BeTrue();
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Commands/GetScanTests.cs:
```diff
@@ -0,0 +1,19 @@
+    var command = new GetScan(ScanId);
+
+    // assert
+    command.Url.Should().Be($"/api/v1/scans/{ScanId}");
+    command.Method.Should().Be(HttpMethod.Get);
+    command.ExpectReply.Should().BeTrue();
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Commands/DeleteScanTests.cs:
```diff
@@ -0,0 +1,19 @@
+    var command = new DeleteScan(ScanId);
+
+    // assert
+    command.Url.Should().Be($"/api/v1/scans/{ScanId}/delete");
+    command.Method.Should().Be(HttpMethod.Get);
+    command.ExpectReply.Should().BeFalse();
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Commands/CreateScanTests.cs:
```diff
@@ -0,0 +1,51 @@
+
+    // assert
+    var content = command.Body is null ? default : await command.Body.ReadAsStringAsync().ConfigureAwait(false);
+    content.Should().Be(expectedContent);
+    command.Url.Should().Be("/api/v1/scans");
+    command.Method.Should().Be(HttpMethod.Post);
+    command.ExpectReply.Should().BeTrue();
```
ditto

ostridm committed:
feat(scan): introduce a client to manage scans via REST API

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032076772
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032076823
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032076870
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032076966
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032101095
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032128899
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032374208
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032376302
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032376358
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032376420
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032376477
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1032376548

closes #44

ostridm committed:
feat(scan): introduce a client to manage scans via REST API

rebase and adjust using directives

closes #44

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
Erm, why do you do this in tests?

derevnjuk commented on test/SecTester.Scan.Tests/Models/UploadHarOptionsTests.cs:
```diff
@@ -1,3 +1,6 @@
+using SecTester.Scan.Models;
+using SecTester.Scan.Target.Har;
```
We are using global using

derevnjuk commented on test/SecTester.Scan.Tests/Target/TargetTests.cs:
```diff
@@ -1,6 +1,7 @@
 using System.Net.Http.Headers;
 using System.Net.Http.Json;
 using System.Text;
+using SecTester.Scan.Target.Har;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Fixtures/ScanFixture.cs:
```diff
@@ -0,0 +1,55 @@
+using NSubstitute.ClearExtensions;
+using SecTester.Core;
+using SecTester.Core.Bus;
+using SecTester.Scan.CI;
+using SecTester.Scan.Models;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -0,0 +1,183 @@
+using SecTester.Core.Exceptions;
+using SecTester.Scan.Commands;
+using SecTester.Scan.Models;
+using SecTester.Scan.Target.Har;
+using SecTester.Scan.Tests.Fixtures;
+using Request = SecTester.Scan.Models.Request;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Commands/CreateScanTests.cs:
```diff
@@ -0,0 +1,60 @@
+using System.Text;
+using SecTester.Bus.Dispatchers;
+using SecTester.Scan.Commands;
+using SecTester.Scan.Tests.Fixtures;
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs:
```diff
@@ -0,0 +1,22 @@
+using SecTester.Scan.CI;
```
ditto

ostridm committed:
feat(scan): introduce a client to manage scans via REST API

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1033322194
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1033322121
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1033321996
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1033321868
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1033321664
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1033321292
https://github.com/NeuraLegion/sectester-net/pull/58#discussion_r1033320078

closes #44

derevnjuk committed:
style(scan): clean up code

closes #44

derevnjuk commented:
@ostridm make sure that husky is working in your case, otherwise, manually format files before pushing. Make sure that you utilize LF instead of CRLF https://github.com/NeuraLegion/sectester-net/pull/58/commits/67cb3da2a34aa9c09ed28bcaa5a7186451ef2113
